File: orbital-back/routers/assets.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List
from database import get_db
from models import Asset, AssetStatus
from schemas import AssetCreate, AssetOut, AssetStats

router = APIRouter(prefix="/api/v1/assets", tags=["assets"])


@router.get("/stats", response_model=AssetStats)
def get_asset_stats(db: Session = Depends(get_db)):
    """Get asset statistics"""
    total = db.query(Asset).count()
    nominal = db.query(Asset).filter(Asset.status == AssetStatus.ACTIVE).count()
    at_risk = db.query(Asset).filter(
        (Asset.status == AssetStatus.WARNING) | (Asset.status == AssetStatus.MAINTENANCE)
    ).count()
    
    return AssetStats(
        total_assets=total,
        nominal_assets=nominal,
        at_risk_assets=at_risk
    )


@router.get("/", response_model=List[AssetOut])
def list_assets(db: Session = Depends(get_db)):
    """List all assets"""
    assets = db.query(Asset).order_by(Asset.created_at.desc()).all()
    return assets


@router.get("/{asset_id}", response_model=AssetOut)
def get_asset(asset_id: str, db: Session = Depends(get_db)):
    """Get a specific asset by asset_id"""
    asset = db.query(Asset).filter(Asset.asset_id == asset_id).first()
    if not asset:
        raise HTTPException(status_code=404, detail="Asset not found")
    return asset


@router.post("/", response_model=AssetOut)
def create_asset(asset_data: AssetCreate, db: Session = Depends(get_db)):
    """Create a new asset"""
    # Check if asset_id already exists
    existing = db.query(Asset).filter(Asset.asset_id == asset_data.asset_id).first()
    if existing:
        raise HTTPException(status_code=400, detail="Asset ID already exists")
    
    # Create new asset
    new_asset = Asset(
        asset_id=asset_data.asset_id,
        asset_type=asset_data.asset_type,
        status=asset_data.status,
        latitude=asset_data.latitude,
        longitude=asset_data.longitude,
        altitude_km=asset_data.altitude_km,
        velocity_x=asset_data.velocity_x,
        velocity_y=asset_data.velocity_y,
        velocity_z=asset_data.velocity_z,
        battery_voltage=asset_data.battery_voltage,
        coordinate_array=asset_data.coordinate_array
    )
    
    db.add(new_asset)
    db.commit()
    db.refresh(new_asset)
    
    return new_asset


@router.delete("/{asset_id}")
def delete_asset(asset_id: str, db: Session = Depends(get_db)):
    """Delete an asset"""
    asset = db.query(Asset).filter(Asset.asset_id == asset_id).first()
    if not asset:
        raise HTTPException(status_code=404, detail="Asset not found")
    
    db.delete(asset)
    db.commit()
    
    return {"message": "Asset deleted successfully"}

File: orbital-front/src/lib/api/assets.ts
const API_BASE = "http://localhost:8000/api/v1";

export interface Asset {
  id: number;
  asset_id: string;
  asset_type: string;
  status: string;
  registration_timestamp: string;
  latitude: number | null;
  longitude: number | null;
  altitude_km: number | null;
  velocity_x: number | null;
  velocity_y: number | null;
  velocity_z: number | null;
  battery_voltage: number | null;
  coordinate_array: string | null;
  created_at: string;
}

export interface AssetStats {
  total_assets: number;
  nominal_assets: number;
  at_risk_assets: number;
}

export interface AssetCreate {
  asset_id: string;
  asset_type: string;
  status: string;
  latitude?: number;
  longitude?: number;
  altitude_km?: number;
  velocity_x?: number;
  velocity_y?: number;
  velocity_z?: number;
  battery_voltage?: number;
  coordinate_array?: string;
}

export async function getAssetStats(): Promise<AssetStats> {
  const response = await fetch(`${API_BASE}/assets/stats`);
  if (!response.ok) throw new Error("Failed to fetch asset stats");
  return response.json();
}

export async function listAssets(): Promise<Asset[]> {
  const response = await fetch(`${API_BASE}/assets/`);
  if (!response.ok) throw new Error("Failed to fetch assets");
  return response.json();
}

export async function getAsset(assetId: string): Promise<Asset> {
  const response = await fetch(`${API_BASE}/assets/${assetId}`);
  if (!response.ok) throw new Error("Failed to fetch asset");
  return response.json();
}

export async function createAsset(data: AssetCreate): Promise<Asset> {
  const response = await fetch(`${API_BASE}/assets/`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || "Failed to create asset");
  }
  return response.json();
}

export async function deleteAsset(assetId: string): Promise<void> {
  const response = await fetch(`${API_BASE}/assets/${assetId}`, {
    method: "DELETE",
  });
  if (!response.ok) throw new Error("Failed to delete asset");
}
